Perf sql dialect

amanPras opened the pull request:
## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] No regressions
- [x] Performance improvement for cobol sources with exec sql block 

## Checklist:
- [ ] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

amanPras committed:
perf: improve db2 expression performance - try 1

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

ap891843 commented:
report from positive Test for the improvement from commit b92ac256f399b8e006baf7f204b70893d613d0c7 [filename.csv](https://github.com/user-attachments/files/18321705/filename.csv)

Stats:
```
positive improvement count     218.000000
negative improvement count      72.000000
positive_mean        7.041687
negative_mean       -6.661591
overall_mean         3.639494
max_improvement     81.604278
min_improvement    -34.537532
```
![image](https://github.com/user-attachments/assets/8483b091-3661-483a-aeb8-f785c47c8429)

amanPras committed:
perf: reduce ambiguity in the dbs expression

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
perf: move semantics checks from parser to visitor

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras commented:
@slavek-kucera @ishche opened this PR for suggestions

ap891843 commented:
Below stats is based on the test performance data generated by pipeline (this branch vs development)

> ** -ve _mean_ better performance as this is calculated using 

`df_merged['Improvement'] = ((df_merged['Implicit dialects processing_optimise'] - df_merged['Implicit dialects processing_development']) / df_merged['Implicit dialects processing_development']) * 100`

positive_count      42.000000
negative_count     248.000000
positive_mean       21.898019
negative_mean      -11.488877
overall_mean        -6.653533
max_improvement     75.556349
min_improvement    -61.447500

amanPras committed:
perf: refactor db2 parser rule expression

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

ap891843 commented:
[result.csv](https://github.com/user-attachments/files/18359728/result.csv)

amanPras committed:
perf: refactor db2 grammar to reduce ambiguities

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
perf: reduce lookahead for column

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
perf: small improvements around table correlation clause

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
perf: reduce ambiguity around fullselect and subselect

reduce ambiguity around fullselect and subselect,  by removing the order_offset_fetch clause
and keeping it part of full select.
This looks to be compatible with what Db2 preprocessor behaviour

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -267,7 +268,7 @@ dbs_alter_trigger_options: ((DISALLOW | ALLOW | DISABLE) DEBUG MODE | QUALIFIER
                             CONCENTRATE STATEMENTS (OFF | WITH LITERALS))*; /*random ordering req */
 dbs_alter_trigger_replace: REPLACE (ACTIVE VERSION | VERSION dbs_trigger_version_id)? dbs_alter_trigger_spec;
-dbs_alter_trigger_spec: ((NO CASCADE)? BEFORE | AFTER | INSTEAD OF) (INSERT | DELETE | UPDATE (OF dbs_column_name (dbs_comma_separator dbs_column_name)*)?) ON (dbs_table_name | dbs_view_name)
+dbs_alter_trigger_spec: ((NO CASCADE)? BEFORE | AFTER | INSTEAD OF) (INSERT | DELETE | UPDATE (OF dbs_column_name (dbs_comma_separator dbs_column_name)*)?) ON (dbs_host_variable | dsb_alias_name)
```
dsb_alias_name - > dbs_alias_name ?

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -315,12 +316,12 @@ dbs_comment: COMMENT ON (dbs_comment_multiple_column_list | (dbs_comment_alias_d
              dbs_comment_role | dbs_comment_sequence | dbs_comment_table | dbs_comment_trigger | dbs_comment_trusted |
              dbs_comment_type | dbs_comment_mask | dbs_comment_permission | dbs_comment_variable) IS dbs_string_constant);
-dbs_comment_multiple_column_list: (dbs_table_name | dbs_view_name) LPARENCHAR dbs_column_name IS
+dbs_comment_multiple_column_list: (dbs_host_variable | dsb_alias_name) LPARENCHAR dbs_column_name IS
```
I am unable to build an example with a host variable.
Also it seems to reject `'quoted table names'`.

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -356,7 +357,7 @@ dbs_create: CREATE (dbs_create_alias | dbs_create_aux_table | dbs_create_db | db
 //CREATE ALIAS
 dbs_create_alias: PUBLIC? ALIAS (table_alias | sequence_alias);
-table_alias: dsb_alias_name FOR TABLE? (dbs_table_name | dbs_view_name | dsb_alias_name);
+table_alias: dsb_alias_name FOR TABLE? (dbs_host_variable | dsb_alias_name);
 sequence_alias: dsb_alias_name FOR SEQUENCE dbs_sequence_name;
```
Could you merge these rules?

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -712,20 +713,22 @@ dbs_explain_stabilized: STABILIZED DYNAMIC QUERY STMTID (NUMERICLITERAL | dbs_in
+//ref: https://www.ibm.com/docs/en/db2-for-zos/13?topic=statements-fetch
+dbs_fetch: FETCH (INSENSITIVE | SENSITIVE)? (WITH CONTINUE)? (BEFORE | AFTER | dbs_fetch_rowpos | dbs_fetch_rowsetpos)?
+            FROM? dbs_cursor_name dbs_fetch_single_multi_row;
+dbs_fetch_single_multi_row: (FOR (dbs_host_variable | dbs_integer_constant) ROWS)?  dbs_fetch_singlerow?;
```
The multiple-row-fetch clause seems to be allowed only when one of the rowset options is specified.

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -786,29 +789,26 @@ dbs_include_sqlca: INCLUDE SQLCA;
 /*INSERT */
-dbs_insert: INSERT INTO (dbs_table_name | dbs_view_name) (LPARENCHAR dbs_column_name (dbs_comma_separator dbs_column_name)* RPARENCHAR)?
+dbs_insert: INSERT INTO (dbs_host_variable | dsb_alias_name) (LPARENCHAR dbs_column_name (dbs_comma_separator dbs_column_name)* RPARENCHAR)?
             dbs_insert_include? (OVERRIDING USER VALUE)? dbs_insert_values;
```
full-select instead of dbs_insert_values is allowed.

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -1744,32 +1748,34 @@ dbs_triggered_sql_statement_adv: dbs_call | dbs_delete | dbs_get_diagnostics_sta
 dbs_collection_id: dbs_sql_identifier;
-dbs_column_name: (dsb_alias_name DOT_FS)?  dbs_sql_identifier; //TODO {validateLength($T.text, "Column name", 30);};
+//dbs_column_name: (dsb_alias_name DOT_FS)?  IDENTIFIER;
+dbs_column_name: dbs_sql_identifier (DOT_FS  (dbs_sql_identifier (DOT_FS dbs_sql_identifier)?)?  dbs_sql_identifier)?; //TODO {validateLength($T.text, "Column name", 30);};
```
dbs_column_name in this form cannot be used in the context of e.g. ALTER TABLE statement.

amanPras committed:
chore: rename dsb_alias_name - > dbs_alias_name

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecValidatorVisitor.java:
```diff
@@ -0,0 +1,221 @@
+
+  @Override
+  public List<Node> visitDbs_decfloat_integer(Db2SqlExecParser.Dbs_decfloat_integerContext ctx) {
+    int value = Integer.parseInt(ctx.getText());
```
Won't this throw on overflow?

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecValidatorVisitor.java:
```diff
@@ -0,0 +1,221 @@
+
+  @Override
+  public List<Node> visitDbs_decimal_15_31(Db2SqlExecParser.Dbs_decimal_15_31Context ctx) {
+    validateTokenWithRegex(ctx, "\\b(15|31)\\b", "15 or 31 are only allowed");
```
In this case, isn't it equivalent to .equals("15") || .equals("31") ?

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecValidatorVisitor.java:
```diff
@@ -0,0 +1,221 @@
+
+  @Override
+  public List<Node> visitDbs_maxPartition(Db2SqlExecParser.Dbs_maxPartitionContext ctx) {
+    int intInputValue = Integer.parseInt(ctx.getText());
```
handle the exception.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecValidatorVisitor.java:
```diff
@@ -0,0 +1,221 @@
+  }
+
+  protected void validateTextInRange(ParserRuleContext ctx, int min, int max) {
+    int value = Integer.parseInt(ctx.getText());
```
exception handling.

amanPras committed:
perf: reduce lookahead for select and select_into

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: changes for validation code as per PR comments

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

perf: update grammar as per the review comments

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: changes for validation code as per PR comments

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
test: update tests which sre not valid as per compiler and latest Cobol LS changes

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/sql/TestSqlInsertStatement.java:
```diff
@@ -141,7 +141,9 @@ class TestSqlInsertStatement {
     return Stream.of(
-        INSERT1, INSERT2, INSERT3, INSERT4, INSERT5, INSERT6, INSERT7, INSERT8, INSERT9, INSERT10);
+        INSERT1, INSERT2, INSERT3, INSERT4, INSERT5,
+            // INSERT6, // Removed as compilation fails for this syntax. Include is not supported by compiler though the doc mentions it
```
This might be the correct usage if the insert include statement (but I am not sure)
```
SELECT * INTO :A, :B, :C FROM FINAL TABLE (
INSERT INTO ABC (C) INCLUDE (A SMALLINT,   
B SMALLINT) VALUES (:TAD+1)                
)                                          
```

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -712,20 +721,22 @@ dbs_explain_stabilized: STABILIZED DYNAMIC QUERY STMTID (NUMERICLITERAL | dbs_in
-dbs_fetch_singlemulti: (INSENSITIVE | SENSITIVE)? (dbs_fetch_single | dbs_fetch_multi);
-dbs_fetch_single: (WITH CONTINUE)? dbs_fetch_rowpos? FROM? dbs_cursor_name dbs_fetch_singlerow?;
+//ref: https://www.ibm.com/docs/en/db2-for-zos/13?topic=statements-fetch
+dbs_fetch: FETCH (BEFORE | AFTER) FROM? dbs_cursor_name
```
I don't think that the before/after is mandatory.

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -786,35 +797,35 @@ dbs_include_sqlca: INCLUDE SQLCA;
-            dbs_insert_include? (OVERRIDING USER VALUE)? dbs_insert_values;
-dbs_insert_include: INCLUDE LPARENCHAR dbs_column_name dbs_include_data_type (dbs_comma_separator dbs_column_name dbs_include_data_type)* RPARENCHAR;
+dbs_insert: INSERT INTO dbs_alias_name (LPARENCHAR dbs_column_name (dbs_comma_separator dbs_column_name)* RPARENCHAR)?
+            dbs_fullselect? (OVERRIDING USER VALUE)? dbs_insert_values;
```
Should this be just the following?
```suggestion
            (OVERRIDING USER VALUE)? dbs_insert_values;
```

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -1743,33 +1763,36 @@ dbs_triggered_sql_statement_adv: dbs_call | dbs_delete | dbs_get_diagnostics_sta
-dbs_column_name: (dsb_alias_name DOT_FS)?  dbs_sql_identifier; //TODO {validateLength($T.text, "Column name", 30);};
+//dbs_column_name: (dsb_alias_name DOT_FS)?  IDENTIFIER;
+//dbs_column_name: dbs_sql_identifier (DOT_FS  (dbs_sql_identifier (DOT_FS dbs_sql_identifier)?)?  dbs_sql_identifier)?; //TODO {validateLength($T.text, "Column name", 30);};
+dbs_column_name: (dbs_alias_name DOT_FS)?  dbs_sql_identifier;
```
This still does not work in the context of `alter table add column`

amanPras replied:
allowed: `alter table add add COLUMN col1 SMALLINT NOT NULL DEFAULT`
Not allowed: `alter table add add COLUMN add.col1 SMALLINT NOT NULL DEFAULT`

amanPras committed:
revert: revert changes for dbs_insert as per doc

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
perf: update dbs_fetch with a prefix

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: update db2 grammar as per PR comments

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
chore: clean up commented code

Signed-off-by: amanprashant <aman.prashant@broadcom.com>
